### PR TITLE
Feat(logger): Add a logger service

### LINKED ISSUE #67: Logger service
We should add a logger service to reduce the console noise.

### PR DESCRIPTION
**Pull request by fsmanuel**
I would propose something like that to close #67
Any feedback is welcome!

### CODE REVIEW

File: app/services/logger.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    'send'
+  ],
+
+  log(type) {
+    if (!get(this, 'allowedTypes').includes(type)) {
+      throw new Error(`You specified a unknown type: "${type}".`);
+    }
+
+    if (get(this, 'showLogsFor').includes(type)) {
+      console.log.apply(null, arguments);
```
Comment (skddc) [inline app/services/logger.js]: I think this should use `Ember.Logger` instead.

File: app/services/smt.js [hunk 2]
```diff
@@ -49,7 +50,7 @@ export default Ember.Service.extend({
       }
     };
 
-    Ember.Logger.debug('connecting to irc', credentials);
+    this.get('logger').log('connection', 'connecting to irc', credentials);
```
Comment (skddc) [inline app/services/smt.js]: Maybe we can mix this in somehow so it's just `this.log()`?

File: app/services/logger.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    get(this, 'showLogsFor').push(type);
+  },
+
+  removeFromLogs(type) {
+    let types = get(this, 'showLogsFor');
+    let index = types.indexOf(type);
+
+    if (index > -1) {
+        types.splice(index, 1);
+    }
```
Comment (galfert) [inline app/services/logger.js]: Couldn't you just use Ember's `removeObject` here?

**Comment by fsmanuel** [comment]
@skddc @galfert updated

File: app/routes/application.js [hunk 4]
```diff
@@ -8,8 +8,16 @@ const {
 } = Ember;
 
 export default Route.extend({
+  logger: service(),
   smt: service(),
 
+  beforeModel() {
```
Comment (galfert) [inline app/routes/application.js]: You should at least call `this._super(...arguments)` here.

Comment (fsmanuel) [inline app/routes/application.js]: Good catch!

Comment (fsmanuel) [inline app/routes/application.js]: updated

File: app/routes/application.js [hunk 5]
```diff
@@ -8,8 +8,17 @@ const {
...
+    this._super(...arguments);
+    // See a list of allowed types in logger.js
+    // Add or remove all your log types here:
+    // this.get('logger').addToLogs('message');
+    // this.get('logger').removeFromLogs('join');
```
Comment (skddc) [inline app/routes/application.js]: What do you think about adding a line that would remove all logs at once, when uncommenting it?

Comment (fsmanuel) [inline app/routes/application.js]: I changed the public api to `.add(type)` `.remove(type)` and added `.disable()` and `.enable()` + tests

**Comment by fsmanuel** [comment]
@skddc @galfert updated

**Comment by skddc** [comment]
Very cool!

I added the remaining log statements (which I stilll saw when logs weren't disabled) and formatted the type in the output with brackets.

I think it's good to merge now.

**Comment by fsmanuel** [comment]
👍

**Comment by galfert** [comment]
Looking good 👍 

Btw, you can access the logger during runtime from the console via

`window.Hyperchannel.__container__.lookup('service:logger')`

**Comment by skddc** [comment]
> Btw, you can access the logger during runtime from the console via
> 
> window.Hyperchannel.__container__.lookup('service:logger')

Nice, thanks!

**Pull request merged**